game/src/ecs/components/ai/fight/RangeAI.java

```diff
@@ -0,0 +1,89 @@
+
+    private final float attackRange;
+    private final float distance;
+    private final Skill skill;
+    boolean isNotBetweenAttackAndDistance = true;
+    private GraphPath<Tile> path;
+
+    /**
+     * Attacks the player if he is within the given range between attackRange and distance. Otherwise, it will move into that range.
+     *
+     * @param attackRange max. Range in which the attack skill should be executed
+     * @param distance min. Range in which the attack skill should be executed
+     * @param skill Skill to be used when an attack is performed
+     */
+
+    public RangeAI(float attackRange, float distance, Skill skill){
+        this.attackRange = attackRange;
+        this.distance = distance;
+        this.skill = skill;
+    }
+
+    @Override
+    public void fight(Entity entity) {
+        if(attackRange > distance) {
+            if (AITools.playerInRange(entity, distance) && isNotBetweenAttackAndDistance) {
+                isNotBetweenAttackAndDistance = false;
+                Point positionHero = getPosition(Game.getHero().get());
+                Point positionEntity = getPosition(entity);
+                List<Tile> tiles = getAccessibleTilesInRange(positionEntity, attackRange - distance);
+                Point newPosition;
+                boolean newPositionFound = false;
+                for (int index = 0; index < tiles.size(); index++) {
+                    newPosition = tiles.get(index).getCoordinate().toPoint();
+                    if (!AITools.inRange(newPosition, positionHero, distance)) {
+                        path = AITools.calculatePath(positionEntity, newPosition);
+                        newPositionFound = true;
+                        break;
+                    }
+                }
+                if (!newPositionFound) {
+                    tiles = getAccessibleTilesInRange(positionHero, distance);
+                    newPosition = tiles.get(random.nextInt(tiles.size())).getCoordinate().toPoint();
+                    path = AITools.calculatePath(positionEntity, newPosition);
+                }
+            }
+
+            if (!AITools.playerInRange(entity, attackRange) && isNotBetweenAttackAndDistance) {
+                isNotBetweenAttackAndDistance = false;
+                path = AITools.calculatePathToHero(entity);
+            }
+
+            if (!(AITools.playerInRange(entity, attackRange) && !AITools.playerInRange(entity, distance))) {
+                AITools.move(entity, path);
+                isNotBetweenAttackAndDistance = true;
+            } else {
+                skill.execute(entity);
+            }
+        }
+    }
+
+    private Point getPosition(Entity entity){
+        return ((PositionComponent)
+            entity.getComponent(PositionComponent.class)
+                .orElseThrow(
+                    () ->
+                        new MissingComponentException(
+                            "PositionComponent")))
```

Comment:
Hier wäre doch toll, wenn ich noch erfahren würde, welche Entität es war z.b. "Hero is missing PositionComponent".
Aber erstmal nur als Gedanke.

Reply:
Habe ich angepasst.